nifi-commons/nifi-utils/src/test/java/org/apache/nifi/util/TestFormatUtils.java

```diff
@@ -0,0 +1,368 @@
+                        (TimeUnit.MINUTES), 7 * 24 * 60L,
+                        (TimeUnit.SECONDS), 7 * 24 * 60 * 60L,
+                        (TimeUnit.MILLISECONDS), 7 * 24 * 60 * 60 * 1000L,
+                        (TimeUnit.MICROSECONDS), 7 * 24 * 60 * 60 * (1000 * 1000L),
+                        (TimeUnit.NANOSECONDS), 7 * 24 * 60 * 60 * (1000 * 1000 * 1000L)
+                ).entrySet().stream()
+                .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue() * 1.5));
+
+        Map<TimeUnit, Double> onePointFiveWeeksInOtherUnits = Arrays.stream(TimeUnit.values())
+                .filter(timeUnit -> !TimeUnit.DAYS.equals(timeUnit))
+                .collect(Collectors.toMap(destinationUnit -> destinationUnit,
+                        destinationUnit -> FormatUtils.getPreciseTimeDuration(ONE_AND_A_HALF_WEEKS, destinationUnit)));
+
+        onePointFiveWeeksInOtherUnits.forEach((key, value) ->
+                assertEquals(ONE_POINT_FIVE_WEEKS_IN_OTHER_UNITS.get(key), value)
+        );
+    }
+
+    /**
+     * Positive flow test for decimal time inputs
+     */
+    @Test
+    public void testGetPreciseTimeDurationShouldHandleDecimalValues() {
+        final List<String> WHOLE_NUMBERS = List.of("10 ms", "10 millis", "10 milliseconds");
+        final List<String> DECIMAL_NUMBERS = List.of("0.010 s", "0.010 seconds");
+        final double EXPECTED_MILLIS = 10.0;
+
+        List<Double> parsedWholeMillis = WHOLE_NUMBERS.stream()
+                .map(whole -> FormatUtils.getPreciseTimeDuration(whole, TimeUnit.MILLISECONDS))
+                .collect(Collectors.toList());
+        List<Double> parsedDecimalMillis = DECIMAL_NUMBERS.stream()
+                .map(decimal -> FormatUtils.getPreciseTimeDuration(decimal, TimeUnit.MILLISECONDS))
+                .collect(Collectors.toList());
+
+        parsedWholeMillis.forEach(it -> assertEquals(EXPECTED_MILLIS, it));
+        parsedDecimalMillis.forEach(it -> assertEquals(EXPECTED_MILLIS, it));
+    }
+
+    /**
+     * Positive flow test for decimal inputs that are extremely small
+     */
+    @Test
+    public void testGetPreciseTimeDurationShouldHandleSmallDecimalValues() {
+        final Map<String, Map<String, Object>> SCENARIOS = Map.of(
+                "decimalNanos", Map.of("originalUnits", TimeUnit.NANOSECONDS, "expectedUnits", TimeUnit.NANOSECONDS, "originalValue", 123.4, "expectedValue", 123.0),
+                "lessThanOneNano", Map.of("originalUnits", TimeUnit.NANOSECONDS, "expectedUnits", TimeUnit.NANOSECONDS, "originalValue", 0.9, "expectedValue", 1.0),
+                "lessThanOneNanoToMillis", Map.of("originalUnits", TimeUnit.NANOSECONDS, "expectedUnits", TimeUnit.MILLISECONDS, "originalValue", 0.9, "expectedValue", 0.0),
+                "decimalMillisToNanos", Map.of("originalUnits", TimeUnit.MILLISECONDS, "expectedUnits", TimeUnit.NANOSECONDS, "originalValue", 123.4, "expectedValue", 123_400_000.0)
```

Comment:
The nested `Map.of()` calls are a bit difficult to read. Also, it would be helpful to backport this change to the 1.x support branch, where Java 8 is still required. With that in mind, it might make more sense to re-evaluate the test method approach to make it more intuitive, as opposed to using a Map of Maps. Using JUnit 5 parameterized tests would be much better.